src/main/scala/uk/gov/nationalarchives/Lambda.scala

```diff
@@ -0,0 +1,85 @@
+package uk.gov.nationalarchives
+
+import cats.effect.IO
+import cats.effect.unsafe.implicits.global
+import com.amazonaws.services.lambda.runtime.{Context, RequestStreamHandler}
+import fs2.interop.reactivestreams._
+import pureconfig.ConfigSource
+import pureconfig.generic.auto._
+import pureconfig.module.catseffect.syntax._
+import software.amazon.awssdk.transfer.s3.model.CompletedUpload
+import uk.gov.nationalarchives.Lambda.{Config, StepFnInput}
+import upickle.default
+import upickle.default._
+
+import java.io.{InputStream, OutputStream}
+import scala.io.Source
+import scala.xml.PrettyPrinter
+
+class Lambda extends RequestStreamHandler {
+  val dAS3Client: DAS3Client[IO] = DAS3Client[IO]()
+  implicit val inputReader: Reader[StepFnInput] = macroR[StepFnInput]
+
+  def handleRequest(input: InputStream, output: OutputStream, context: Context): Unit = {
+    val rawInput: String = Source.fromInputStream(input).mkString
+    val stepFnInput = default.read[StepFnInput](rawInput)
+    val keyPrefix = s"opex/${stepFnInput.executionId}/"
+    val opexFileName = s"$keyPrefix${stepFnInput.executionId}.opex"
+
+    for {
+      config <- ConfigSource.default.loadF[IO, Config]()
+      publisher <- dAS3Client.listCommonPrefixes(config.stagingCacheBucket, keyPrefix)
+      completedUpload <- publisher
+        .toStreamBuffered[IO](1024 * 5)
+        .through(accumulatePrefixes)
+        .map(generateOpexWithManifest)
+        .flatMap { opexXmlString => uploadToS3(opexXmlString, opexFileName, config.stagingCacheBucket) }
+        .compile
+        .toList
+      verifiedCompletedUpload <-
```

Comment:
You can simplify this a bit using the newly found `raiseWhen`
```scala
        .compile
        .toList
      _ <- IO.raiseWhen(completedUpload.isEmpty)(new Exception(s"No uploads were attempted for '$keyPrefix'"))
    } yield completedUpload.head
```

Reply:
👍 . That's looks nicer.